core/dbt/contracts/graph/nodes.py

```diff
@@ -608,32 +648,96 @@ def same_contract(self, old) -> bool:
+        from dbt.adapters.factory import get_adapter_constraint_support
+        from dbt.adapters.base import ConstraintSupport
+
+        constraint_support = get_adapter_constraint_support(adapter_type)
+
         # Next, compare each column from the previous contract (old.columns)
-        for key, value in sorted(old.columns.items()):
+        for old_key, old_value in sorted(old.columns.items()):
             # Has this column been removed?
-            if key not in self.columns.keys():
-                columns_removed.append(value.name)
+            if old_key not in self.columns.keys():
+                columns_removed.append(old_value.name)
             # Has this column's data type changed?
-            elif value.data_type != self.columns[key].data_type:
+            elif old_value.data_type != self.columns[old_key].data_type:
                 column_type_changes.append(
-                    (str(value.name), str(value.data_type), str(self.columns[key].data_type))
+                    (
+                        str(old_value.name),
+                        str(old_value.data_type),
+                        str(self.columns[old_key].data_type),
+                    )
                 )
 
+            # Have enforced columns level constraints changed?
+            # Constraints are only enforced for table and incremental materializations.
+            # We only really care if the old node was one of those materializations for breaking changes
+            if (
+                old_key in self.columns.keys()
+                and old_value.constraints != self.columns[old_key].constraints
+                and old.config.materialized in ["table", "incremental"]
+            ):
+
+                for old_constraint in old_value.constraints:
+                    if (
+                        old_constraint not in self.columns[old_key].constraints
+                        and constraint_support[old_constraint.type] == ConstraintSupport.ENFORCED
+                    ):
+                        enforced_column_constraint_removed.append(
+                            [old_key, str(old_constraint.type)]
+                        )
+
+        # Now compare the model level constraints
+        if old.constraints != self.constraints and old.config.materialized in [
+            "table",
+            "incremental",
+        ]:
+            for old_constraint in old.constraints:
+                if (
+                    old_constraint not in self.constraints
+                    and constraint_support[old_constraint.type] == ConstraintSupport.ENFORCED
+                ):
+                    enforced_model_constraint_removed.append(
+                        (str(old_constraint.type), old_constraint.columns)
+                    )
+
+        # Check for materialization changes.
+        if (
+            old.config.materialized in ["table", "incremental"]
+            and old.config.materialized != self.config.materialized
```

Comment:
should changing between table and incremental materializations actually raise a breaking contract error? I can see how going from a table -> view is effectively like deleting any existing constraints but going from table -> incremental should maintain any enforced constraints.

Reply:
along the same lines - we should probably only raise a breaking contract error if the materialization changes to one where constraints are unenforceable if were any enforceable (by the adapter) constraints defined to begin with. 

dbt can still validate contracts (column names + types) in materializations where constraints are unsupported. so if you move from a table -> view and there were no enforceable constraints to begin with or breaking changes to the contract, no error should be raised.

Reply:
That makes sense.  Updated.